README.md
# Concentrate

A packet arbiter.

## Code Status

[![Build Status](https://travis-ci.com/helium/concentrate.svg?token=gSksRQHcDis4sPKF5NRm&branch=master)](https://travis-ci.com/helium/concentrate)

## Required Toolchain

You will need to download and install the ''aarch64-linux-gnu-gcc' toolchain and add the 'aarch64-unknown-linux-gnu' target to rustup in order to build this application for the current target (Rasberry Pi 3 B+).

The included .cargo/config file will specify the linker needed to complete the build. Setup instructions by system are as follows:

### Linux
Install the toolchain to your system:
#### Arch
Install the toolchain:
```sh
> sudo pacman -S aarch64-linux-gnu-gcc
```
#### Ubuntu
```sh
> sudo apt-get install gcc-aarch64-linux-gnu
```
#### Common
Add it to Rust:
```sh
> rustup target add aarch64-unknown-linux-gnu
```
Now building is easy:
```sh
> cargo build --target aarch64-unknown-linux-gnu [--release]
```
## Testing
If you are putting the binary on an otherwise working Raspberry Pi image, you will want to kill the pre-existing concentrate service:
```sh
sudo sv d /etc/sv/concentrate
```
If you would like to run a test session with a _remote_ client, you can start a concentrate server with the following options:
```sh
sudo concentrate -p -r 192.168.1.xxx serve
```
On the remote machine, such a local development machine, you can run the client:
```sh
concentrate -p listen
```

### Notes

For 10.76.100.10: 
```
./concentrate serve -p --publish 10.76.100.1:31337 --listen 0.0.0.0:31338

CONCENTRATE_LOG="debug" ./target/debug/concentrate longfi --radio-listen 10.76.100.10:31338 --radio-publish 0.0.0.0:31337 --longfi-listen 127.0.0.1:31341 --longfi-publish 127.0.0.1:31340

./target/debug/concentrate longfi-test --publish 127.0.0.1:31340 --listen 127.0.0.1:31341
```

For 10.76.100.11:
```
./concentrate serve -p --publish 10.76.100.1:31330 --listen 0.0.0.0:31331

CONCENTRATE_LOG="debug" ./target/debug/concentrate longfi --radio-listen 10.76.100.11:31331 --radio-publish 0.0.0.0:31330 --longfi-listen 127.0.0.1:31332 --longfi-publish 127.0.0.1:31333

./target/debug/concentrate longfi-test --publish 127.0.0.1:31333 --listen 127.0.0.1:31332
```


CHANGELOG.md
<!--
M.m.p (YYYY-MM-DD)
==================
Add a summary of this release.

**BREAKING CHANGES**:

* Some change which breaks API or ABI compatiblity with.


Feature enhancements:

* [Link to github PR]():
  A new feature.

Bug fixes:

* [Link to github PR]():
  A bugfix.
-->

0.2.9 (2019-11-27)
==================
Feature enhancements:

* [#48](https://github.com/helium/concentrate/pull/48):
  Allow runtime choice of logging backend.

0.2.8 (2019-11-26)
==================
Housekeeping:

* [#47](https://github.com/helium/concentrate/pull/47):
  Add SX1302 support.

0.2.7 (2019-11-18)
==================
Feature enhancements:

* [#46](https://github.com/helium/concentrate/pull/46):
  Add connect `subcommand`.

0.2.6 (2019-11-08)
==================
Feature enhancements:

* [#45](https://github.com/helium/concentrate/pull/45):
  Add tag bits to proto message.

0.2.4 (2019-10-01)
==================
Bug fixes:

* [#36](https://github.com/helium/concentrate/pull/36):
  Take timestamp, snr, rssi from first fragment.

Feature enhancements:

* [#37](https://github.com/helium/concentrate/pull/37):
  Make `concentrate` executable the root level project.

* [#38](https://github.com/helium/concentrate/pull/38):
  Upgrade to the 2018 edition.

0.2.3 (2019-08-08)
==================
Bug fixes:

* [#34](https://github.com/helium/concentrate/pull/34):
  Default poll interval changed from 100ms to 10ms.
  Epoll triggering mode set to level instead of edge.

0.2.2 (2019-08-08)
==================
Bug fixes:

* [#32](https://github.com/helium/concentrate/pull/32):
  Drop CRC errored fragments from reassembly

0.2.1 (2019-08-07)
==================
Bug fixes:

* [#30](https://github.com/helium/concentrate/pull/30):
  Remove LongFi parsing timeouts

0.2.0 (2019-08-06)
==================
Bug fixes:

* [#29](https://github.com/helium/concentrate/pull/29):
  Fix LongFi [de]fragmentation.

0.1.9 (2019-07-28)
==================
Feature enhancements:

* [#26](https://github.com/helium/concentrate/pull/26):
  Add built in self test.

* [#23](https://github.com/helium/concentrate/pull/23):
  LongFi App for Concentrate.

0.1.8 (2019-07-23)
==================
Bug fixes:

* [#18](https://github.com/helium/concentrate/pull/18):
  Make calls to `libloragw-sys` threadsafe.

* [#21](https://github.com/helium/concentrate/pull/21):
  Add `.cargo/config` for cross-compilation. Extend `serve`/`listen` to allow `serve` to run on hardware but listener to be remote; `remote-ip` cli option was added to enable this.

* [#24](https://github.com/helium/concentrate/pull/24):
  Block in `transmit()` until Concentrator is able to accept a new TX packet.

0.1.7 (2019-06-26)
==================
Bug fixes:

* [#17](https://github.com/helium/concentrate/pull/17):
  Add lockfile to allow building with outdated Buildroot.

0.1.6 (2019-06-25)
==================
Feature enhancements:

* [#16](https://github.com/helium/concentrate/pull/16):
  Update reset line to use `/dev/gpio` symlink for portability.

0.1.5 (2019-06-21)
==================
Feature enhancements:

* [#15](https://github.com/helium/concentrate/pull/15):
  Add `syslog` logging backend.

0.1.4 (2019-06-19)
==================
Feature enhancements:

* [#14](https://github.com/helium/concentrate/pull/14):
  Switch to synchronous (req/resp) network interface.

0.1.3 (2019-06-17)
==================
Feature enhancements:

* [#11](https://github.com/helium/concentrate/pull/11):
  Allow enginering/scientific notation for frequency arguments.
* [#10](https://github.com/helium/concentrate/pull/10):
  Add implicit header flag to `send` command.

Bug fixes:

* [#13](https://github.com/helium/concentrate/pull/13):
  Lower SPI clock speed as it might be contributing to TX errors.

0.1.2 (2019-06-07)
==================
Feature enhancements:

* [#8](https://github.com/helium/concentrate/pull/8):
  Update channel scheme to be outside of LoRaWan.
* [#7](https://github.com/helium/concentrate/pull/7):
  Add transmit gain lookup-table to configuration.

0.1.1 (2019-06-03)
==================
Feature enhancements:

* [#4](https://github.com/helium/concentrate/pull/4):
  Improved error reporting on failed FFI calls.

Bug fixes:

* [#5](https://github.com/helium/concentrate/pull/5):
  Update RSSI offset per RAK2245/2247 spec.


src/cmdline.rs
use std::{net::SocketAddr, path::PathBuf};
use structopt::{clap, StructOpt};

#[derive(Debug, StructOpt)]
pub enum Cmd {
    /// Built In Self Test. Configures, starts, then immediately stops
    /// concentrator. Returns 1 on hardware failure.
    #[structopt(name = "bist", raw(setting = "clap::AppSettings::Hidden"))]
    Bist,

    /// Operate as a consumer of another instance running as the
    /// server. This mode is primarily meant for debugging and
    /// printing [de]serialized packets.
    #[structopt(name = "listen")]
    Listen(Listen),

    /// Operate as a LongFi server. Requires separate `concentrate
    /// serve` instance.
    #[structopt(name = "longfi")]
    LongFi(LongFi),

    /// This mode is primarily meant for testing and debugging LongFi
    /// device code.
    #[structopt(name = "longfi-test", raw(setting = "clap::AppSettings::Hidden"))]
    LongFiTest(LongFiTest),

    /// Transmit a packet using provided string as payload. Similar to
    /// 'listen', requires another instance running in 'serve' mode.
    #[structopt(name = "send")]
    Send(Send),

    /// Operate as a server between concentrator hardware and UDP
    /// clients.
    #[structopt(name = "serve")]
    Serve(Serve),

    /// Check connection to concentrator chip and immediately exit
    /// with status code.
    #[structopt(name = "connect")]
    Connect,
}

#[derive(Debug, StructOpt)]
pub struct Listen {
    /// Print packets. `-p` will print on a single line, and `-pp`
    /// will pretty-print over several.
    #[structopt(short = "p", parse(from_occurrences))]
    pub print_level: u8,

    /// Address on which to listen for received uplink packets.
    #[structopt(
        value_name = "ADDR",
        short = "l",
        long = "listen",
        default_value = "127.0.0.1:31337"
    )]
    pub listen_addr_in: SocketAddr,
}

#[derive(Debug, StructOpt)]
pub struct LongFi {
    /// Address on which to listen for to-be-transmitted downlink
    /// packets.
    #[structopt(
        value_name = "ADDR",
        long = "longfi-listen",
        default_value = "127.0.0.1:31341"
    )]
    pub longfi_listen_addr_in: SocketAddr,

    /// Address to publish received LongFi packets to.
    #[structopt(
        value_name = "ADDR",
        long = "longfi-publish",
        default_value = "127.0.0.1:31340"
    )]
    pub longfi_publish_addr_out: SocketAddr,

    /// Address to publish raw to-be-transmitted LoRa packets to.
    #[structopt(
        value_name = "ADDR",
        long = "radio-listen",
        default_value = "127.0.0.1:31338"
    )]
    pub radio_listen_addr_out: SocketAddr,

    /// Address on which to listen for raw uplink packets.
    #[structopt(
        value_name = "ADDR",
        long = "radio-publish",
        default_value = "127.0.0.1:31337"
    )]
    pub radio_publish_addr_in: SocketAddr,
}

#[derive(Debug, StructOpt)]
pub struct LongFiTest {
    /// Address to send LongFi requests to.
    #[structopt(
        value_name = "ADDR",
        long = "listen",
        default_value = "127.0.0.1:31341"
    )]
    pub listen_addr_out: SocketAddr,

    /// Address on which to listen for LongFi responses.
    #[structopt(
        value_name = "ADDR",
        long = "publish",
        default_value = "127.0.0.1:31340"
    )]
    pub publish_addr_in: SocketAddr,

    /// number of bytes to transmit
    #[structopt(
        value_name = "NUM_BYTES",
        short = "n",
        long = "num-bytes",
        default_value = "90"
    )]
    pub num_bytes: usize,

    /// disable fragmentation, making it so that bytes are all send monolithically
    #[structopt(short = "d", long = "disable-frag")]
    pub disable_fragmentation: bool,
}

#[derive(Debug, StructOpt)]
pub struct Send {
    /// Address to send raw LoRa packets to.
    #[structopt(
        value_name = "ADDR",
        long = "listen",
        default_value = "127.0.0.1:31338"
    )]
    pub listen_addr_out: SocketAddr,

    /// Address on which to listen for transmit responses.
    #[structopt(
        value_name = "ADDR",
        long = "publish",
        default_value = "127.0.0.1:31337"
    )]
    pub publish_addr_in: SocketAddr,

    /// Print packets. `-p` will print on a single line, and `-pp`
    /// will pretty-print over several.
    #[structopt(short = "p", parse(from_occurrences))]
    pub print_level: u8,

    /// Transmit with implicit header.
    #[structopt(short = "i", long = "implicit")]
    pub implicit: bool,

    /// Frequency to transmit packet on.
    #[structopt(value_name = "Hz", short = "f", long = "freq")]
    pub freq: f64,

    /// Radio [0,1] to transmit on.
    #[structopt(short = "r", long = "radio")]
    pub radio: u8,

    /// Transmit power.
    #[structopt(value_name = "dBm", short = "o", long = "power", default_value = "0")]
    pub power: i8,

    /// Spreading factor [7,8,9,10,11,12].
    #[structopt(
        value_name = "SF",
        short = "s",
        long = "spreading",
        default_value = "7"
    )]
    pub spreading: u8,

    /// Coderate [5,6,7,8]. Actual coderate is 4/[VALUE].
    #[structopt(short = "c", long = "coderate", default_value = "5")]
    pub coderate: u8,

    /// Bandwidth [7800,15600,31200,62500,125000,250000,500000]
    #[structopt(
        value_name = "Hz",
        short = "b",
        long = "bandwidth",
        default_value = "125000"
    )]
    pub bandwidth: u32,

    /// String payload.
    pub payload: Option<String>,
}

#[derive(Debug, StructOpt)]
pub struct Serve {
    /// Polling interval. How often to poll concentrator's FIFO for
    /// received packets.
    #[structopt(
        value_name = "MILLISECONDS",
        short = "I",
        long = "interval",
        default_value = "10"
    )]
    pub interval: u64,

    /// Print packets. `-p` will print on a single line, and `-pp`
    /// will pretty-print over several.
    #[structopt(short = "p", parse(from_occurrences))]
    pub print_level: u8,

    #[structopt(short = "c", long = "config", parse(from_os_str))]
    pub cfg_file: Option<PathBuf>,

    /// Address on which to listen for requests.
    #[structopt(
        value_name = "ADDR",
        short = "l",
        long = "listen",
        default_value = "127.0.0.1:31338"
    )]
    pub listen_addr_in: SocketAddr,

    /// Address to send responses and received uplink packets to.
    #[structopt(
        value_name = "ADDR",
        short = "u",
        long = "publish",
        default_value = "127.0.0.1:31337"
    )]
    pub publish_addr_out: SocketAddr,
}
